### ISSUE TITLE
don't fold(..) an Option when you can map_or(..)

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#1658]
catch `my_option.iter().fold(my_val, |x, y| ..y)`, suggest `my_option.map_or(my_val, |x| ..my_val)`

**Comment by CYBAI** [comment]
Hi @llogiq ,

I'd like to work on this but I'm a little confused by the example code

```rust
let opt = Some(234);

let my_val = 10;
let a = opt.iter().fold(my_val, |x, y| ..y);
let b = opt.map_or(my_val, |x| ..my_val);

println!("{:?}, {:?}", a, b);
```

When I ran above code, it will throw the mismatch type error

```rust
error[E0308]: mismatched types
 --> src/main.rs:5:44
  |
5 |     let a = opt.iter().fold(my_val, |x, y| ..y);
  |                                            ^^^ expected integral variable, found struct `std::ops::RangeTo`
  |
  = note: expected type `{integer}`
             found type `std::ops::RangeTo<&{integer}>`

error[E0308]: mismatched types
 --> src/main.rs:6:36
  |
6 |     let b = opt.map_or(my_val, |x| ..my_val);
  |                                    ^^^^^^^^ expected integral variable, found struct `std::ops::RangeTo`
  |
  = note: expected type `{integer}`
             found type `std::ops::RangeTo<{integer}>`

error: aborting due to 2 previous errors
```

BTW, should the new lint include the case @dwijnand mentioned in #3009 ?

```rust
let as_is = match forgot_ty.ty_adt_def() {
  Some(def) => def.has_dtor(cx.tcx),
  _ => false
};

let to_be = forgot_ty.ty_adt_def().map_or(false, |def| def.has_dtor(cx.tcx));
```

**Comment by flip1995** [comment]
Here's some working code that displays the thought of this lint: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=72d9d5ed7a5492b065d71199edcb0957)

So if you fold an `Option` and the `Option` is `None` then the initial value is returned (empty iterator). But this is the same as calling `map_or` without an iterator.

What's fixed in #3009 is a different case. The lint described here is 
```rust
opt.iter().fold(init, |acc, y| do_something_with(acc, y))
vs
opt.map_or(init, |y| do_something_with(init, y))
```
The other case is `match opt { Some(y) => do_somthing_with(init, y), _ => init }` vs `opt.map_or(init, |y| do_something_with(init, y))`

The former probably belongs in the `methods` module, and the later in the `matches` module IMO.

**Comment by clueless-skywatcher** [comment]
Hi! Can I pick this up?
@rustbot claim

**Comment by clueless-skywatcher** [comment]
My bad, didn’t notice the merged PR.

**Comment by bugprone** [comment]
@rustbot claim

Note for anyone following: the merged PR mentioned above (#3009, 2018) covers a different case, as flip1995 noted at the time — current clippy (1.96) emits no warning on `opt.iter().fold(init, |acc, x| ...)`, and master only has `unnecessary_fold`/`manual_try_fold`, neither of which handles this. Planning to extend `unnecessary_fold` with an Option-receiver case suggesting `as_ref().map_or(...)`/`map_or(...)` rather than adding a new lint.
